File: packages/react-native-video/ios/hybrids/VideoPlayerEmitter/HybridVideoPlayerEventEmitter.swift
//
//  HybridVideoPlayerEventEmitter.swift
//  ReactNativeVideo
//
//  Created by Krzysztof Moch on 02/05/2025.
//

import Foundation
import NitroModules

struct ListenerPair {
  let id: UUID
  let eventName: String
  let callback: Any
}

class HybridVideoPlayerEventEmitter: HybridVideoPlayerEventEmitterSpec {
  var listeners: [ListenerPair] = []

  // MARK: - Private helpers

  private func addListener<T>(eventName: String, listener: T) -> ListenerSubscription {
    let id = UUID()
    listeners.append(ListenerPair(id: id, eventName: eventName, callback: listener))
    return ListenerSubscription(remove: { [weak self] in
      self?.listeners.removeAll { $0.id == id }
    })
  }

  private func emitEvent<T>(eventName: String, invoke: (T) throws -> Void) {
    for pair in listeners where pair.eventName == eventName {
      if let callback = pair.callback as? T {
        do {
          try invoke(callback)
        } catch {
          print("[ReactNativeVideo] Error calling \(eventName) listener: \(error)")
        }
      } else {
        print("[ReactNativeVideo] Invalid callback type for \(eventName)")
      }
    }
  }

  // MARK: - Listener registration methods

  func addOnAudioBecomingNoisyListener(listener: @escaping () -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onAudioBecomingNoisy", listener: listener)
  }

  func addOnAudioFocusChangeListener(listener: @escaping (Bool) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onAudioFocusChange", listener: listener)
  }

  func addOnBandwidthUpdateListener(listener: @escaping (BandwidthData) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onBandwidthUpdate", listener: listener)
  }

  func addOnBufferListener(listener: @escaping (Bool) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onBuffer", listener: listener)
  }

  func addOnControlsVisibleChangeListener(listener: @escaping (Bool) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onControlsVisibleChange", listener: listener)
  }

  func addOnEndListener(listener: @escaping () -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onEnd", listener: listener)
  }

  func addOnExternalPlaybackChangeListener(listener: @escaping (Bool) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onExternalPlaybackChange", listener: listener)
  }

  func addOnLoadListener(listener: @escaping (onLoadData) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onLoad", listener: listener)
  }

  func addOnLoadStartListener(listener: @escaping (onLoadStartData) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onLoadStart", listener: listener)
  }

  func addOnPlaybackStateChangeListener(listener: @escaping (onPlaybackStateChangeData) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onPlaybackStateChange", listener: listener)
  }

  func addOnPlaybackRateChangeListener(listener: @escaping (Double) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onPlaybackRateChange", listener: listener)
  }

  func addOnProgressListener(listener: @escaping (onProgressData) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onProgress", listener: listener)
  }

  func addOnReadyToDisplayListener(listener: @escaping () -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onReadyToDisplay", listener: listener)
  }

  func addOnSeekListener(listener: @escaping (Double) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onSeek", listener: listener)
  }

  func addOnStatusChangeListener(listener: @escaping (VideoPlayerStatus) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onStatusChange", listener: listener)
  }

  func addOnTimedMetadataListener(listener: @escaping (TimedMetadata) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onTimedMetadata", listener: listener)
  }

  func addOnTextTrackDataChangedListener(listener: @escaping ([String]) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onTextTrackDataChanged", listener: listener)
  }

  func addOnTrackChangeListener(listener: @escaping (Variant_NullType_TextTrack?) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onTrackChange", listener: listener)
  }

  func addOnVolumeChangeListener(listener: @escaping (onVolumeChangeData) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onVolumeChange", listener: listener)
  }

  func clearAllListeners() throws {
    listeners.removeAll()
  }

  // MARK: - Event emission methods

  func onAudioBecomingNoisy() {
    emitEvent(eventName: "onAudioBecomingNoisy") { (callback: () throws -> Void) in try callback() }
  }

  func onAudioFocusChange(_ hasFocus: Bool) {
    emitEvent(eventName: "onAudioFocusChange") { (callback: (Bool) throws -> Void) in try callback(hasFocus) }
  }

  func onBandwidthUpdate(_ data: BandwidthData) {
    emitEvent(eventName: "onBandwidthUpdate") { (callback: (BandwidthData) throws -> Void) in try callback(data) }
  }

  func onBuffer(_ isBuffering: Bool) {
    emitEvent(eventName: "onBuffer") { (callback: (Bool) throws -> Void) in try callback(isBuffering) }
  }

  func onControlsVisibleChange(_ isVisible: Bool) {
    emitEvent(eventName: "onControlsVisibleChange") { (callback: (Bool) throws -> Void) in try callback(isVisible) }
  }

  func onEnd() {
    emitEvent(eventName: "onEnd") { (callback: () throws -> Void) in try callback() }
  }

  func onExternalPlaybackChange(_ isExternalPlaybackActive: Bool) {
    emitEvent(eventName: "onExternalPlaybackChange") { (callback: (Bool) throws -> Void) in try callback(isExternalPlaybackActive) }
  }

  func onLoad(_ data: onLoadData) {
    emitEvent(eventName: "onLoad") { (callback: (onLoadData) throws -> Void) in try callback(data) }
  }

  func onLoadStart(_ data: onLoadStartData) {
    emitEvent(eventName: "onLoadStart") { (callback: (onLoadStartData) throws -> Void) in try callback(data) }
  }

  func onPlaybackStateChange(_ data: onPlaybackStateChangeData) {
    emitEvent(eventName: "onPlaybackStateChange") { (callback: (onPlaybackStateChangeData) throws -> Void) in try callback(data) }
  }

  func onPlaybackRateChange(_ rate: Double) {
    emitEvent(eventName: "onPlaybackRateChange") { (callback: (Double) throws -> Void) in try callback(rate) }
  }

  func onProgress(_ data: onProgressData) {
    emitEvent(eventName: "onProgress") { (callback: (onProgressData) throws -> Void) in try callback(data) }
  }

  func onReadyToDisplay() {
    emitEvent(eventName: "onReadyToDisplay") { (callback: () throws -> Void) in try callback() }
  }

  func onSeek(_ position: Double) {
    emitEvent(eventName: "onSeek") { (callback: (Double) throws -> Void) in try callback(position) }
  }

  func onStatusChange(_ status: VideoPlayerStatus) {
    emitEvent(eventName: "onStatusChange") { (callback: (VideoPlayerStatus) throws -> Void) in try callback(status) }
  }

  func onTimedMetadata(_ metadata: TimedMetadata) {
    emitEvent(eventName: "onTimedMetadata") { (callback: (TimedMetadata) throws -> Void) in try callback(metadata) }
  }

  func onTextTrackDataChanged(_ tracks: [String]) {
    emitEvent(eventName: "onTextTrackDataChanged") { (callback: ([String]) throws -> Void) in try callback(tracks) }
  }

  func onTrackChange(_ track: Variant_NullType_TextTrack?) {
    emitEvent(eventName: "onTrackChange") { (callback: (Variant_NullType_TextTrack?) throws -> Void) in try callback(track) }
  }

  func onVolumeChange(_ data: onVolumeChangeData) {
    emitEvent(eventName: "onVolumeChange") { (callback: (onVolumeChangeData) throws -> Void) in try callback(data) }
  }
}

File: packages/react-native-video/android/src/main/java/com/twg/video/hybrids/videoplayereventemitter/HybridVideoPlayerEventEmitter.kt
package com.margelo.nitro.video

import android.util.Log
import com.margelo.nitro.core.NullType
import java.util.UUID

data class ListenerPair(val id: UUID, val eventName: String, val callback: Any)

class HybridVideoPlayerEventEmitter : HybridVideoPlayerEventEmitterSpec() {
  private val lock = Any()

  var listeners: MutableList<ListenerPair> = mutableListOf()

  // MARK: - Private helpers
  private fun <T : Any> addListener(eventName: String, listener: T): ListenerSubscription {
    val id = UUID.randomUUID()
    synchronized(lock) {
      listeners.add(ListenerPair(id, eventName, listener))
    }
    return ListenerSubscription {
      synchronized(lock) {
        listeners.removeAll { it.id == id }
      }
    }
  }

  private inline fun <reified T> emitEvent(eventName: String, invokeCallback: (T) -> Unit) {
    val snapshot: List<ListenerPair> = synchronized(lock) {
      listeners.filter { it.eventName == eventName }.toList()
    }

    snapshot.forEach { pair ->
      try {
        @Suppress("UNCHECKED_CAST")
        val callback = pair.callback as? T ?: run {
          Log.d(TAG, "Invalid callback type for $eventName")
          return@forEach
        }
        invokeCallback(callback)
      } catch (t: Throwable) {
        Log.d(TAG, "Error calling $eventName listener", t)
      }
    }
  }


  // MARK: - Listener registration methods
  
  override fun addOnAudioBecomingNoisyListener(listener: () -> Unit) =
    addListener("onAudioBecomingNoisy", listener)

  override fun addOnAudioFocusChangeListener(listener: (Boolean) -> Unit) =
    addListener("onAudioFocusChange", listener)

  override fun addOnBandwidthUpdateListener(listener: (BandwidthData) -> Unit) =
    addListener("onBandwidthUpdate", listener)

  override fun addOnBufferListener(listener: (Boolean) -> Unit) =
    addListener("onBuffer", listener)

  override fun addOnControlsVisibleChangeListener(listener: (Boolean) -> Unit) =
    addListener("onControlsVisibleChange", listener)

  override fun addOnEndListener(listener: () -> Unit) =
    addListener("onEnd", listener)

  override fun addOnExternalPlaybackChangeListener(listener: (Boolean) -> Unit) =
    addListener("onExternalPlaybackChange", listener)

  override fun addOnLoadListener(listener: (onLoadData) -> Unit) =
    addListener("onLoad", listener)

  override fun addOnLoadStartListener(listener: (onLoadStartData) -> Unit) =
    addListener("onLoadStart", listener)

  override fun addOnPlaybackStateChangeListener(listener: (onPlaybackStateChangeData) -> Unit) =
    addListener("onPlaybackStateChange", listener)

  override fun addOnPlaybackRateChangeListener(listener: (Double) -> Unit) =
    addListener("onPlaybackRateChange", listener)

  override fun addOnProgressListener(listener: (onProgressData) -> Unit) =
    addListener("onProgress", listener)

  override fun addOnReadyToDisplayListener(listener: () -> Unit) =
    addListener("onReadyToDisplay", listener)

  override fun addOnSeekListener(listener: (Double) -> Unit) =
    addListener("onSeek", listener)

  override fun addOnStatusChangeListener(listener: (VideoPlayerStatus) -> Unit) =
    addListener("onStatusChange", listener)

  override fun addOnTimedMetadataListener(listener: (TimedMetadata) -> Unit) =
    addListener("onTimedMetadata", listener)

  override fun addOnTextTrackDataChangedListener(listener: (Array<String>) -> Unit) =
    addListener("onTextTrackDataChanged", listener)

  override fun addOnTrackChangeListener(listener: (Variant_NullType_TextTrack?) -> Unit) =
    addListener("onTrackChange", listener)

  override fun addOnVolumeChangeListener(listener: (onVolumeChangeData) -> Unit) =
    addListener("onVolumeChange", listener)

  override fun addOnAdsResolvedListener(listener: (AdsResolvedEvent) -> Unit) =
    addListener("onAdsResolved", listener)

  override fun addOnAdBreakStartListener(listener: (AdBreakEvent) -> Unit) =
    addListener("onAdBreakStart", listener)

  override fun addOnAdBreakEndListener(listener: (AdBreakEvent) -> Unit) =
    addListener("onAdBreakEnd", listener)

  override fun addOnAdProgressListener(listener: (AdProgressInfo) -> Unit) =
    addListener("onAdProgress", listener)

  override fun addOnAdStartListener(listener: (AdInfo) -> Unit) =
    addListener("onAdStart", listener)

  override fun addOnAdCompleteListener(listener: (AdInfo) -> Unit) =
    addListener("onAdComplete", listener)

  override fun addOnAdSkippedListener(listener: (AdInfo) -> Unit) =
    addListener("onAdSkipped", listener)

  override fun addOnAdClickedListener(listener: () -> Unit) =
    addListener("onAdClicked", listener)

  override fun addOnAdErrorListener(listener: (AdErrorEvent) -> Unit) =
    addListener("onAdError", listener)

  override fun addOnAllAdsCompletedListener(listener: () -> Unit) =
    addListener("onAllAdsCompleted", listener)

  override fun addOnAdStateChangeListener(listener: (VideoAdState) -> Unit) =
    addListener("onAdStateChange", listener)

  override fun clearAllListeners() {
    synchronized(lock) {
      listeners.clear()
    }
  }

  // MARK: - Event emission methods

  fun onAudioBecomingNoisy() =
    emitEvent<() -> Unit>("onAudioBecomingNoisy") { it() }

  fun onAudioFocusChange(hasFocus: Boolean) =
    emitEvent<(Boolean) -> Unit>("onAudioFocusChange") { it(hasFocus) }

  fun onBandwidthUpdate(data: BandwidthData) =
    emitEvent<(BandwidthData) -> Unit>("onBandwidthUpdate") { it(data) }

  fun onBuffer(isBuffering: Boolean) =
    emitEvent<(Boolean) -> Unit>("onBuffer") { it(isBuffering) }

  fun onControlsVisibleChange(isVisible: Boolean) =
    emitEvent<(Boolean) -> Unit>("onControlsVisibleChange") { it(isVisible) }

  fun onEnd() =
    emitEvent<() -> Unit>("onEnd") { it() }

  fun onExternalPlaybackChange(isExternalPlayback: Boolean) =
    emitEvent<(Boolean) -> Unit>("onExternalPlaybackChange") { it(isExternalPlayback) }

  fun onLoad(data: onLoadData) =
    emitEvent<(onLoadData) -> Unit>("onLoad") { it(data) }

  fun onLoadStart(data: onLoadStartData) =
    emitEvent<(onLoadStartData) -> Unit>("onLoadStart") { it(data) }

  fun onPlaybackStateChange(data: onPlaybackStateChangeData) =
    emitEvent<(onPlaybackStateChangeData) -> Unit>("onPlaybackStateChange") { it(data) }

  fun onPlaybackRateChange(rate: Double) =
    emitEvent<(Double) -> Unit>("onPlaybackRateChange") { it(rate) }

  fun onProgress(data: onProgressData) =
    emitEvent<(onProgressData) -> Unit>("onProgress") { it(data) }

  fun onReadyToDisplay() =
    emitEvent<() -> Unit>("onReadyToDisplay") { it() }

  fun onSeek(position: Double) =
    emitEvent<(Double) -> Unit>("onSeek") { it(position) }

  fun onTimedMetadata(metadata: TimedMetadata) =
    emitEvent<(TimedMetadata) -> Unit>("onTimedMetadata") { it(metadata) }

  fun onTextTrackDataChanged(tracks: Array<String>) =
    emitEvent<(Array<String>) -> Unit>("onTextTrackDataChanged") { it(tracks) }

  fun onTrackChange(track: TextTrack?) {
    val param = if (track == null) {
      Variant_NullType_TextTrack.create(NullType.NULL)
    } else {
      Variant_NullType_TextTrack.create(track)
    }
    emitEvent<(Variant_NullType_TextTrack?) -> Unit>("onTrackChange") { it(param) }
  }

  fun onVolumeChange(data: onVolumeChangeData) =
    emitEvent<(onVolumeChangeData) -> Unit>("onVolumeChange") { it(data) }

  fun onStatusChange(status: VideoPlayerStatus) =
    emitEvent<(VideoPlayerStatus) -> Unit>("onStatusChange") { it(status) }

  fun onAdsResolved(data: AdsResolvedEvent) =
    emitEvent<(AdsResolvedEvent) -> Unit>("onAdsResolved") { it(data) }

  fun onAdBreakStart(data: AdBreakEvent) =
    emitEvent<(AdBreakEvent) -> Unit>("onAdBreakStart") { it(data) }

  fun onAdBreakEnd(data: AdBreakEvent) =
    emitEvent<(AdBreakEvent) -> Unit>("onAdBreakEnd") { it(data) }

  fun onAdProgress(data: AdProgressInfo) =
    emitEvent<(AdProgressInfo) -> Unit>("onAdProgress") { it(data) }

  fun onAdStart(data: AdInfo) =
    emitEvent<(AdInfo) -> Unit>("onAdStart") { it(data) }

  fun onAdComplete(data: AdInfo) =
    emitEvent<(AdInfo) -> Unit>("onAdComplete") { it(data) }

  fun onAdSkipped(data: AdInfo) =
    emitEvent<(AdInfo) -> Unit>("onAdSkipped") { it(data) }

  fun onAdClicked() =
    emitEvent<() -> Unit>("onAdClicked") { it() }

  fun onAdError(data: AdErrorEvent) =
    emitEvent<(AdErrorEvent) -> Unit>("onAdError") { it(data) }

  fun onAllAdsCompleted() =
    emitEvent<() -> Unit>("onAllAdsCompleted") { it() }

  fun onAdStateChange(state: VideoAdState) =
    emitEvent<(VideoAdState) -> Unit>("onAdStateChange") { it(state) }

  companion object {
    const val TAG = "HybridVideoPlayerEventEmitter"
  }
}
